src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs

```diff
@@ -396,6 +395,7 @@ public void LinkPrefix(BlobBuilder prefix)
             // avoid chaining empty chunks:
             if (prefix.Count == 0)
             {
+                prefix.ClearAndFreeChunk();
                 return;
```

Comment:
Not sure what is the correct fix, the empty chunk is not added into the chain so I assume will not be serialized either. Looks there is no way to free it from the parent/head. If the fix is only about freeing the initial state this fix might be enough, note that the empty `BlobBuilder` stays writeable after the link. @jaredpar is this fix enough for your scenario? 

@pedrobsaila could you add some test?

Reply:
I didn't add a test because I didn't find an easy way to check if a chunk is freed or not

Reply:
> Not sure what is the correct fix, the empty chunk is not added into the chain so I assume will not be serialized either.

The expectation when calling `LinkPrefix / LinkSuffix` is that ownership (in terms of pooling) is passing to the receiver. In the case of a zero length `BlobBuilder` there is no ownership transfer, instead the value is dropped. That instead should be returned to the pool aka cleared and freed.

Yes this does produce a slight breaking change in that previously you used to be able to do the following: 

```csharp
var builder1 = new BlobBuilder();
var builder2 = new BlobBuilder();
builder1.LinkSuffix(builder2);
builder2.WriteBytes(0, 1); 
```

At the same time this only works when `blobBuilder2.Length == 0`, in all other cases it would throw. It would be very odd for a consumer to have this pattern at all because `LinkPrefix/LinkSuffix` normally results in the target being read only.

Reply:
@tmat to spot check my assertions here.

Reply:
IsHead is private, can I use reflection to get it ?

Reply:
Think I would make it internal so it could be checked in unit tests

Reply:
@jaredpar Sounds right. At this point I'd say you have more understanding of pooling requirements than I remember from my work on this code many years ago.